kernel/src/grouping_iterator.rs

```diff
@@ -0,0 +1,402 @@
+// This assuming is very important, if the input is not sorted by version, the grouping will not be correct
+impl<L: AsUrl> Iterator for VersionGroupingIterator<L> {
+    type Item = VersionGroup<L>;
+
+    fn next(&mut self) -> Option<VersionGroup<L>> {
+        while let Some(logpath) = self.files.next() {
+            let version: u64 = logpath.version;
+            let mut files = vec![logpath];
+            // this is where we look ahead for the next file and check if it has the same version
+            // if it does, we add it to the current group
+            // if it doesn't, we return the current group and start a new one
+            // this is why we need to assume that the input iterator is already sorted by version, because we only check the next file
+            while let Some(parsed_logpath) = self.files.next_if(|v| v.version == version) {
+                files.push(parsed_logpath)
+            }
+            return Some(VersionGroup { version, files });
+        }
+        None
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::path::PathBuf;
+    use url::Url;
+    /// Returns a URL pointing to the test data directory containing Delta table log files.
+    /// The path is relative to the project root and points to a small test Delta table.
+    /// 
+    /// # Returns
+    /// A URL object representing the canonicalized path to the test Delta log directory
+    fn table_log_dir_url() -> Url {
+        let path = PathBuf::from("./tests/data/table-with-dv-small/_delta_log/");
+        let path = std::fs::canonicalize(path).unwrap();
+        Url::from_directory_path(path).unwrap()
+    }
+
+    /// Creates a ParsedLogPath for testing by constructing a Delta log file path with the given version and type.
+    /// 
+    /// # Arguments
+    /// * `version` - The version number to use in the filename (will be zero-padded to 20 digits)
+    /// * `file_type` - The type of log file to create. Valid options are:
+    ///   * "commit" - Creates a commit file (.json)
+    ///   * "checkpoint" - Creates a single checkpoint file (.checkpoint.parquet)
+    ///   * "multipart1" - Creates part 1 of a 2-part checkpoint
+    ///   * "multipart2" - Creates part 2 of a 2-part checkpoint
+    ///
+    /// # Returns
+    /// A ParsedLogPath containing the constructed URL and parsed metadata
+    ///
+    /// # Panics
+    /// Panics if an invalid file_type is provided
+    fn create_log_path(version: u64, file_type: &str) -> ParsedLogPath<Url> {
+        let base_url = table_log_dir_url();
+        let filename = match file_type {
+            "commit" => format!("{:020}.json", version),
+            "checkpoint" => format!("{:020}.checkpoint.parquet", version),
+            "multipart1" => format!("{:020}.checkpoint.0000000001.0000000002.parquet", version),
+            "multipart2" => format!("{:020}.checkpoint.0000000002.0000000002.parquet", version),
```

Comment:
Maybe just pass in a `LogPathFileType` directly, since each variant contains exactly the necessary info?

@zachschuermann -- it seems like we recently encountered another place that would have benefitted from a utility that creates log path file names like this? Does that ring any bells?